### ISSUE TITLE
Error when calculating gradient of intermediate variable in a layer

### ISSUE THREAD

**Issue by wqhIris** [NVIDIA/apex#1110]
When I modified my code to deploy apex, there throws an error 'DistributedDataParallel' object has no attribute 'callback_queued'.

My related codes that lead to this error are as follow:
![image](https://user-images.githubusercontent.com/33059293/118622234-b4068d00-b7f9-11eb-8db1-13be80864683.png)

My error information is as follow:
![image](https://user-images.githubusercontent.com/33059293/118622572-06e04480-b7fa-11eb-8c1a-2af4c3cbd5cb.png)

I was wondering what should I modify?

**Comment by mindmapper15** [comment]
zero_grad() should be called from optimizer, not from model
Is decoder the optimizer or model?

**Comment by wqhIris** [comment]
> zero_grad() should be called from optimizer, not from model
> Is decoder the optimizer or model?

Thanks for your reply.
The decoder is a model, an additional auxiliary module for generating noise, need not update parameters, but need calculate gradient.
When I use a single gpu, there is no error thrown, and I find there is also zero_grad() defined for model in [Pytorch](https://pytorch.org/docs/stable/generated/torch.nn.Module.html?highlight=zero_grad#torch.nn.Module.zero_grad) .

By the way, I found out that I forget to say, it is calling backward() by adv_distance that causes this error：'DistributedDataParallel' object has no attribute 'callback_queued'. But there is no optimizer for decoder,so I cannot use amp.scale_loss(loss, optimizer) to replace it.

Do you have any advices to solve this error?

**Comment by mindmapper15** [comment]
Sorry for late reply, but this might help to solve the problem.
https://github.com/SeanNaren/deepspeech.pytorch/issues/475

**Comment by wqhIris** [comment]
Hi @mindmapper15, Thanks for your help.
I read the given link, but it is not similar to my situation.
I also try decoder.module.zero_grad(), but I found the error is thrown before the statement "decoder.zero_grad()", and it is thrown at the statement "adv_distance.backward()".

I guess it is due to I did not replace "adv_distance.backward()" with the one following the rule of the "apex". But the "decoder" need not update its parameters, has no optimizer, so I cannot use "with amp.scale_loss(loss, optimizer)". Are there any other modifications?

Thank you again for your attention for this matter.

**Comment by mindmapper15** [comment]
Hello @wqhIris , it's been a long time :)
Have you seen this issue?
https://github.com/NVIDIA/apex/issues/107

**Comment by wqhIris** [comment]
Hi @mindmapper15, 

Thanks a lot for your help and patience,
I have seen issue #107, it seems that in this situation, I need to use DistributedDataParallel in pytorch rather than in apex. I tried 'torch.nn.parallel.DistributedDataParallel', and this error is solved.
